packages/vm/src/imports.rs

```diff
@@ -370,14 +373,14 @@ pub fn do_bls12_381_pairing_equality<
     let s = read_region(&memory, s_ptr, BLS12_381_G2_POINT_LEN)?;
 
     let estimated_point_count = (ps.len() / BLS12_381_G1_POINT_LEN) as u64;
-    let additional_cost = data
-        .gas_config
-        .bls12_381_aggregated_pairing_equality_cost_per_pair
-        // Add one since we do not include any pairs in the base benchmark, and we always need to add one for the `r` and `s` pair.
-        * (estimated_point_count  + 1);
```

Comment:
Is this still the case with the way of calculating the cost I described above on the `LinearGasCost` docs? I took the timings of the `bls12_381_pairing_equality_*` benchmarks and fitted those to a linear equation, which results in the gas costs above.
But adding 1 here makes this more expensive than actually measured, or is this covering some edge case?

Reply:
what we calculate here is the following:

`e(p_1, q_1) × e(p_2, q_2) × … × e(p_n, q_n) = e(s, q)`

which is `n` pairings (`estimated_point_count`) on the left and `k = n+1` terms in total.

This is implemented by moving the pairing to the left

```
e(p_1, q_1) × e(p_2, q_2) × … × e(p_n, q_n) × e(-s, q) = 1
```

giving us `k` terms.

Now lower level all those pairings in this form can be implemented by using the miller loop with `k` entries and a single final exponentiation.

So I think the linear cost model makes sense but with the baseline gas being
- negate s
- final exponentation
and then we have `k = n+1` items with one miller loop operation each.

This is then the same the [proposed Ethereum implementation](https://eips.ethereum.org/EIPS/eip-2537) does
> The cost of the pairing operation is 43000*k + 65000 where k is a number of pairs.

If we measure something that is clearly different, we are doing something wrong. Maybe confuse `n` and `k` somewhere?

Reply:
If I understand correctly, the benchmark is numbered by `n`, so `bls12_381_pairing_equality_8` would be calculating `e(p_1, q_1) × … × e(p_8, q_8) × e(-s, q) = 1`. So, since I just fitted a line through the points `(n, time_of_bls12_381_pairing_equality_n)`, the `e(-s, q)` is currently part of the baseline of the numbers.

I think we could either adjust the benchmark name to `bls12_381_pairing_equality_{num_points+1}` to make more clear that this is the benchmark for an input size of `num_points+1` or I just add a comment above the `bls12_381_pairing_equality_cost` to explain how the benchmark results need to be interpreted / used to calculate the `LinearGasCost`.

Reply:
Does this help? https://github.com/CosmWasm/cosmwasm/pull/2154

Reply:
Yes, that helps! I reran the pairing equality benchmarks with your code and updated the numbers.
Unsurprisingly, this lowers the base cost by the cost of rougly one pairing.

Reply:
Does this PR already include the updated values?